tests/solana_tests/primitives.rs

```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
+
+        contract.constructor("test", &[]);
+        let return_value = contract.function(
+            "mul",
+            &[
+                ethabi::Token::Int(bigint_to_eth(
+                    &first_operand_rand,
+                    width.try_into().unwrap(),
+                )),
+                ethabi::Token::Int(bigint_to_eth(&second_op, width.try_into().unwrap())),
+            ],
+            &[],
+            None,
+        );
+
+        let res = first_operand_rand.mul(second_op);
+        assert_eq!(
+            return_value,
+            vec![ethabi::Token::Int(bigint_to_eth(
+                &res,
+                width.try_into().unwrap(),
+            ))]
+        );
+    }
+}
+
+#[test]
+
+fn test_mul_within_range() {
+    let mut rng = rand::thread_rng();
+    for width in (72..=256).step_by(8) {
+        let src = r#"
+        contract test {
+         
+            function mul(uintN a, uintN b) public returns (uintN) {
+                
+                return a * b;
+                
+            }
+
+           
+
+        }"#
+        .replace("intN", &format!("int{}", width));
+
+        let mut contract = build_solidity(&src);
+        contract.constructor("test", &[]);
+        for _ in 0..10 {
+            // Generate the max value to fit in N bits
+            let limit = BigUint::from(2_u32).pow(width).sub(1_u32);
+
+            // Generate a random number within the the range 0..2^N
+            let first_operand_rand = rng.gen_biguint(width.into());
+
+            // Calculate a number that when multiplied by first_operand_rand, the result will overflow N bits
+            let second_operand_rand = limit.div(&first_operand_rand);
+
+            let return_value = contract.function(
+                "mul",
+                &[
+                    ethabi::Token::Uint(U256::from_big_endian(&first_operand_rand.to_bytes_be())),
+                    ethabi::Token::Uint(U256::from_big_endian(&second_operand_rand.to_bytes_be())),
+                ],
+                &[],
+                None,
+            );
+            let res = first_operand_rand * second_operand_rand;
+
+            assert_eq!(
+                return_value,
+                vec![ethabi::Token::Uint(U256::from_big_endian(
+                    &res.to_bytes_be()
+                ))]
+            );
+        }
+    }
+}
+
+#[test]
+
+fn test_overflow_detect_signed() {
+    let mut rng = rand::thread_rng();
+    for width in (72..=256).step_by(8) {
+        let src = r#"
+        contract test {
+         
+            function mul(intN a, intN b) public returns (intN) {
+                
+                return a * b;
+                
+            }
+
+           
+
+        }"#
+        .replace("intN", &format!("int{}", width));
+        let mut contract = build_solidity(&src);
+
+        contract.constructor("test", &[]);
+
+        // Generate the max value to fit in N bits
+        let limit = BigInt::from(2_u32).pow(width - 1).add(1_u32);
```

Comment:
The bounds for positive and negative numbers are not equal. If your integer is 32-bit wide, for instance, the accepted range of integers is `[-2^31, 2^31-1]`. I would suggest you to add a test to test the boundaries of each case. This would be:

- Test that a result of `-2^31` does not overflow, but a result of `-2^31-1` do overflow.
- Test that a result of `2^31` does not overflow, but a result of `2^31` do overflow.

I used 32 bits as an example here. Please, use the bit widths that are convenient for your use-case.